tests/MongoDB.Driver.Tests/MongoCollectionImplTests.cs

```diff
@@ -258,6 +258,88 @@ public void Settings_should_be_set()
             findOperation.Sort.Should().BeNull();
         }
 
+        [Theory]
+        [ParameterAttributeData]
+        public void Aggregate_should_recognize_merge_collection_argument(
+            [Values(
+                "{ $merge : \"outputcollection\" }",
+                "{ $merge : { into : \"outputcollection\" } }",
+                "{ $merge : { into : { coll : \"outputcollection\" } } }",
+                "{ $merge : { into : { db: \"outputdatabase\", coll : \"outputcollection\" } } }")]
+            string stageDefinitionString,
+            [Values(false, true)] bool async)
+        {
+            var subject = CreateSubject<BsonDocument>();
+            CollectionNamespace expectedCollectionNamespace;
+
+            var stageDefinition = BsonDocument.Parse(stageDefinitionString);
+            if (stageDefinition[0].IsString)
+            {
+                expectedCollectionNamespace = new CollectionNamespace(
+                    subject.CollectionNamespace.DatabaseNamespace,
+                    stageDefinition[0].AsString);
+            }
+            else
+            {
+                var into = stageDefinition[0].AsBsonDocument["into"];
+                if (into.IsString)
+                {
+                    expectedCollectionNamespace = new CollectionNamespace(
+                        subject.CollectionNamespace.DatabaseNamespace,
+                        into.AsString);
+                }
+                else
+                {
+                    if (into.AsBsonDocument.Contains("db"))
+                    {
+                        expectedCollectionNamespace = new CollectionNamespace(
+                            into["db"].AsString,
+                            into["coll"].AsString);
+                    }
+                    else
+                    {
+                        expectedCollectionNamespace = new CollectionNamespace(
+                            subject.CollectionNamespace.DatabaseNamespace,
+                            into["coll"].AsString);
+                    }
+                }
+            }
```

Comment:
As a general rule I don't like to see the same logic IN the test as is being tested (i.e. lines 276-306).

Consider changing the test to remove lines 276-306:

    [Theory]
    [InlineData("{ $merge : \"outputcollection\" }", null, "outputcollection", false)]
    [InlineData("{ $merge : \"outputcollection\" }", null, "outputcollection", true)]
    [InlineData("{ $merge : { into : \"outputcollection\" } }", null, "outputcollection", false)]
    [InlineData("{ $merge : { into : \"outputcollection\" } }", null, "outputcollection", true)]
    [InlineData("{ $merge : { into : { coll : \"outputcollection\" } } }", null, "outputcollection", false)]
    [InlineData("{ $merge : { into : { coll : \"outputcollection\" } } }", null, "outputcollection", true)]
    [InlineData("{ $merge : { into : { db: \"outputdatabase\", coll : \"outputcollection\" } } }", "outputdatabase", "outputcollection", false)]
    [InlineData("{ $merge : { into : { db: \"outputdatabase\", coll : \"outputcollection\" } } }", "outputdatabase", "outputcollection", true)]
    public void Aggregate_should_recognize_merge_collection_argument(
        string stageDefinitionString,
        string expectedDatabaseName,
        string expectedCollectionName,
        bool async)
    {
        var subject = CreateSubject<BsonDocument>();
        var expectedDatabaseNamespace = expectedDatabaseName == null ? subject.Database.DatabaseNamespace : new DatabaseNamespace(expectedDatabaseName);
        var expectedCollectionNamespace = new CollectionNamespace(expectedDatabaseNamespace, expectedCollectionName);

Reply:
Thanks for suggestion!
Done.

Note: I merged database namespace definition into collection namespace creation.